### PR TITLE
C++: Initial AST-based flow through fields

### PR DESCRIPTION
**Pull request by jbj**
This PR is a cleaned-up version of Pavel's branch for adding flow through fields to the AST-based C++ data flow library. I'd like to get this initial code merged so subsequent PRs can be smaller and easier to review.

All the new flow I've seen so far has been beneficial, but I haven't done very extensive tests. I didn't notice any performance issues on Wireshark, but performance could be very dependent on the exact combination of snapshot and query.

Follow-up PRs will be needed to better support flow through nested structs (see `complex.cpp`) and constructors (see `constructors.cpp`).

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -9,10 +9,10 @@ private import semmle.code.cpp.models.interfaces.DataFlow
 cached
 private newtype TNode =
   TExprNode(Expr e) or
-  TParameterNode(Parameter p) { exists(p.getFunction().getBlock()) } or
-  TDefinitionByReferenceNode(VariableAccess va, Expr argument) {
-    definitionByReference(va, argument)
-  } or
+  TPartialDefinitionNode(PartialDefinition pd) or
+  TPostConstructorCallNode(ConstructorCall call) or
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll]: The implicit node associated with a constructor should be pre-update with its post-update node being the corresponding TExprNode.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -5,6 +5,10 @@ private import DataFlowDispatch
 /** Gets the instance argument of a non-static call. */
 private Node getInstanceArgument(Call call) {
   result.asExpr() = call.getQualifier()
+  or
+  // For constructors, there is no qualifier, so we pretend the call itself
+  // is the instance argument.
+  result.asExpr() = call.(ConstructorCall)
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowPrivate.qll]: This should rather be an implicit pre-update node associated with the constructor call.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll [hunk 3]
```diff
@@ -38,11 +38,23 @@ class Node extends TNode {
...
 
   /** Gets the argument that defines this `DefinitionByReferenceNode`, if any. */
   Expr asDefiningArgument() { result = this.(DefinitionByReferenceNode).getArgument() }
 
+  /**
+   * Gets the expression that is partially defined by this node, if any.
+   *
+   * Partial definitions are created for field stores (`x.y = taint();` is a partial
+   * definition of `x`), and for calls that may change the value of an object (so
+   * `x.set(taint())` is a partial definition of `x`, annd `transfer(&x, taint())` is
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll]: typo.
also typo on the line below

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll [hunk 4]
```diff
@@ -177,13 +215,39 @@ class UninitializedNode extends Node, TUninitializedNode {
...
+  PartialDefinitionNode() { this = TPartialDefinitionNode(pd) }
+
+  override Node getPreUpdateNode() { result.asExpr() = pd.getDefinedExpr() }
+
+  override Location getLocation() { result = pd.getLocation() }
+
+  PartialDefinition getPartialDefinition() { result = pd }
+}
+
+class PostConstructorCallNode extends PostUpdateNode, TPostConstructorCallNode {
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll]: This should be changed to extend the extent of `PostUpdateNode` with the `TExprNode` associated with the constructor call.

**Comment by jbj** [comment]
Performance of `FlowVar` on Postgres has gone from bad to worse with this change. I'll try to fix the underlying problem.

**Comment by geoffw0** [comment]
> I didn't notice any performance issues on Wireshark

OverflowStatic.ql, AllocaInLoop.ql, NonConstantFormat.ql, UnsafeCreateProcessCall.ql on cppcheck
took a total of 159s before, now 188s (18% slower)
has removed 7 false positive results for `NonConstantFormat.ql`

OverflowStatic.ql, AllocaInLoop.ql, NonConstantFormat.ql, UnsafeCreateProcessCall.ql on MongoDB
took a total of 231s before, now 319s (38% slower)
has removed 8 false positive results for `NonConstantFormat.ql` (and preserved one possibly correct result)

This isn't very conclusive, but suggests a moderate slowdown is occurring (mostly seen the first and third queries, but that's probably because the other two ran largely on cached data?).  Happy with the results so far but I need to test something with more of them.

**Comment by jbj** [comment]
@geoffw0 Thanks for investigating performance. I've just pushed a commit to restore the lost performance on Postgres. How does that affect your snapshots?

**Comment by geoffw0** [comment]
OverflowDestination.ql, LossyFunctionResultCast.ql, UnsafeDaclSecurityDescriptor.ql, SqlTainted.ql on httpd - took a total of 157s before, now 137s (13% faster)
no results for any of the queries

OverflowDestination.ql, LossyFunctionResultCast.ql, UnsafeDaclSecurityDescriptor.ql, SqlTainted.ql on mysql - took a total of 544s before, now 691s (27% slower)
results are the same

Hmm, I've just noticed that some of the queries (OverflowDestination.ql, SqlTainted.ql) I've been testing use the old security taint library, which is irrelevant to these changes.

The longest run (naively) was LossyFunctionResultCast.ql on mysql.  From a clean cache this took 346s before and 491s after (42% more), explained largely by the top result:
```
LossyFunctionResultCast.ql-14:FlowVar::FlowVar_internal::getAReachedBlockVarSBB#ff ................................. 45.9s (executed 1074 times)
```
->
```
LossyFunctionResultCast.ql-14:FlowVar::FlowVar_internal::getAReachedBlockVarSBB#ff ................................. 2m6s (executed 1170 times)
```
The relation seems to be a fair bit bigger now but I've not investigated any further.

**Comment by geoffw0** [comment]
Trying your new version shortly...

**Comment by geoffw0** [comment]
LossyFunctionResultCast.ql on mysql - 481s, i.e. still running just as slowly:
```
LossyFunctionResultCast.ql-14:FlowVar::FlowVar_internal::getAReachedBlockVarSBB#ff ................................. 1m48s (executed 1167 times)
```
OverflowStatic.ql, AllocaInLoop.ql, NonConstantFormat.ql, UnsafeCreateProcessCall.ql on MongoDB now takes a total 365s, which is no improvement either.  So I think something else must be going on with these other snapshots.

The snapshots I'm using are from here by the way: https://jenkins.internal.semmle.com/job/Language-Tests/job/CPP-HeaderQueryTests/

**Comment by jbj** [comment]
I can reproduce the slowdown on MySQL. It has the same underlying cause as the PostgreSQL slowdown: both projects have a huge Yacc-generated C file that can parse their dialect of SQL. In both projects, the main parser loop (function `yyparse`) is around 20,000 lines long. It looks like this:

```c
yynewstate:
  ⋮
  switch (yyn) {
  case 2: { … } break;
  ⋮
  case 39585: { … } break;
  }
  ⋮
  goto yynewstate;
```

Each of the thousands of switch cases will read or write a couple of variables or fields. A write in one switch case will reach all the ~17,000 `SubBasicBlock`s in the loop.

The liveness analysis I've added on this branch should prevent local vars inside the cases from blowing up, but the problem remains for fields and structs. It gets worse with this PR because `x.a` is now a partial definition of `x`. Maybe I could change it so that's only the case if `x.a` is somehow assigned to. I'll try that.

For the record, I investigated the problems by doing quick-eval of the following `tmp` predicate, pasted into `FlowVar.qll`:

```
  private int getVarTuples(Variable v, int defs, int blocks) {
    defs = strictcount(SubBasicBlock sbb | exists(TBlockVar(sbb, v))) and
    blocks = strictcount(getAReachedBlockVarSBB(TBlockVar(_, v))) and
    result = strictcount(SubBasicBlock def, SubBasicBlock use | use = getAReachedBlockVarSBB(TBlockVar(def, v)))
  }

  /**
   * Holds if `topVar` is one of the variables that contribute the most tuples
   * to `getAReachedBlockVarSBB`. The three integer columns describe how many
   * tuples it contributes, how many definitions it has, and how many unique
   * blocks are reached by those definitions.
   */
  private predicate tmp(Variable topVar, int tuples, int defs, int blocks) {
    topVar = rank[[1..30]](Variable v, int t | t = getVarTuples(v, _, _) | v order by t desc) and
    tuples = getVarTuples(topVar, defs, blocks)
  }
```

**Comment by jbj** [comment]
I looked at the code, and I think read-only `x.a` are handled just fine -- see the `TExplicitFieldStoreQualifier` constructor.

One way to avoid this def-use explosion is to compute SSA instead of def-use. Unfortunately, that would require a complete redesign of `FlowVar`. I think we should rather accept that `FlowVar` is not great and look to IR-based data flow as the long-term solution.

**Comment by jbj** [comment]
The performance problem in MongoDB is caused by the same pattern of code, but this time it's in the PCRE library. It has an interpreter for executing its internal language that looks like

```c
for (;;) {
  ⋮
  switch (opcode) {
  case OP_1:
    ⋮
    return …;
  case OP_2:
    … if (…) continue; …
    return …;
  ⋮
}
```

The interpreter loop is over 5,000 lines long and appears to be hand-written. There are 25 `continue` statements in the main loop, and that's enough to mean that some variables are defined in ~10,000 `SubBasicBlock`s.

The code in `FlowVar` has always been quadratic in the worst case, but these new partial definitions are clearly making the worst case more common. The parser/interpreter loops we've seen here are perhaps pathological, but even straight-line code such as the following is now quadratic because partial defs of qualifiers isn't blocked by other partial defs:

```c
x.a = 1;
x.a = 2;
⋮
x.a = N;
f(x);
```

**Comment by jbj** [comment]
I believe I've addressed all comments now. I'll run a dist build to see if performance is good enough.

**Comment by aschackmull** [comment]
LGTM.

**Comment by jbj** [comment]
I've run a dist build with this change (https://lgtm.com/admin/dist_builds/?create_time=1565690908657&had_default_dist=true), and only four projects timed out. These projects were already close to the limit of 14400 seconds, so either they just got an unlucky worker or we'll just have to accept losing four projects. I think that's a small price to pay for a much-improved analysis.

**Comment by geoffw0** [comment]
Great, I'll repeat the tests I ran before to be safe.  Then if there are no objections I think it's time to merge this.

**Comment by geoffw0** [comment]
OverflowStatic.ql, AllocaInLoop.ql, NonConstantFormat.ql, UnsafeCreateProcessCall.ql on cppcheck
171s (no results), 8% slower than before (the original version of this PR was 18% slower)

OverflowStatic.ql, AllocaInLoop.ql, NonConstantFormat.ql, UnsafeCreateProcessCall.ql on MongoDB
232s (1 result), about the same as before (the original version of this PR was 38% slower)

OverflowDestination.ql, LossyFunctionResultCast.ql, UnsafeDaclSecurityDescriptor.ql, SqlTainted.ql on httpd
142s (no results), 10% faster than before

OverflowDestination.ql, LossyFunctionResultCast.ql, UnsafeDaclSecurityDescriptor.ql, SqlTainted.ql on mysql
618s (2, 3, 0, 1 results), 14% slower than before (the original version of this PR was 27% slower)

LossyFunctionResultCast.ql on mysql
406s (3 results), 17% slower than before (the original version of this PR was 42% slower)
The worst offender on this run is unchanged, except that it's back to taking about half as long:
```
LossyFunctionResultCast.ql-14:FlowVar::FlowVar_internal::getAReachedBlockVarSBB#ff ................................. 54.1s (executed 1071 times)
```


So overall I'm still seing some slowdown but it's substantially reduced and, to my mind, a reasonable tradeoff for this significant feature.  I'm also reassured by your tests, especially on LGTM.

**Comment by jbj** [comment]
Thanks for testing again. I think the slow-downs are acceptable.

**Pull request merged**